### PR TITLE
Add exact source-span lookup to unified search

### LINKED ISSUE #742: Add exact source-span lookup to unified search
## Goal

Let an agent retrieve a bounded, exact source span from a repository file when a compiler, test runner, or other tool has already supplied a file and line location, without falling back to `sed`, Grep, or a general file-read tool.

## Context

PR #740 exposed two different RNA gaps:

1. RNA did not model Rust struct-literal construction sites; #741 tracks that semantic graph gap.
2. After `cargo check` supplied nine exact file/line locations, RNA still could not return those source spans. Targeted source-line reads were required.

The second gap applies to any syntax RNA does not yet model. Exact-span retrieval should complement semantic indexing rather than require every compiler-addressable expression to become a graph node first.

RNA already centralizes code and graph discovery in the unified `search` interface and shares behavior through the service layer. Keep that single surface and make source retrieval explicitly bounded, root-scoped, and safe for non-git directories.

## Acceptance Criteria

- [x] Unified search accepts an exact repo/root-relative file plus a line or bounded line range and returns numbered source text from the current working tree.
- [x] File-and-line lookup works when no indexed symbol begins on that line; it is a source-span operation, not a symbol-search approximation.
- [x] The response identifies repository/root provenance and makes clear that the text is current filesystem state rather than necessarily the last indexed snapshot.
- [x] Ambiguous file matches are rejected with deterministic candidate paths; line lookup never silently chooses the first substring match.
- [x] Paths outside the selected root, parent traversal, and symlink escapes are rejected.
- [x] Line ranges have a documented hard maximum and deterministic truncation/error behavior, preventing the feature from becoming an unbounded file dump.
- [x] Missing files, deleted files, directories, binary/non-UTF-8 files, and out-of-range lines return actionable diagnostics.
- [x] The shared service path delivers identical behavior to CLI and MCP callers; do not create a second MCP-only file-reading implementation.
- [x] Tests cover a compiler-style `path:line:column` location, a span containing syntax with no graph node, non-git roots, ambiguous paths, traversal attempts, and the range bound.
- [x] A real MCP client can retrieve the exact construction-site lines reported by a fixture compiler diagnostic without Grep/Read fallback.

## Trade-offs Accepted

- This is bounded source retrieval, not a replacement for #741's semantic construction-site edges.
- Require exact or uniquely resolved paths when line parameters are present; relevance ranking must not decide which file to read.
- Read the current filesystem so the result reflects uncommitted compiler errors, while labeling that provenance honestly.

## Risk Retirement

The adversarial test must fail implementations that allow `../` or symlink escape, choose one of multiple suffix matches, or return an entire large source file when given a small diagnostic span.

## Notes

- Complementary issue: #741.
- Originating friction: PR #740.
- Likely surfaces: unified search parameters/schema, shared search service, MCP/CLI rendering, root resolution, and bounded filesystem reads.

### PR DESCRIPTION
**Pull request by muness**
Closes #742

## Aim
Agents retrieve exact compiler-reported source spans through RNA instead of falling back to generic file-read tools.

## Problem
Unified search can find indexed semantic nodes, but cannot return a bounded current-working-tree source span for an exact file and line when no symbol begins there. That forces agents to leave RNA precisely when compiler or test diagnostics already provide deterministic locations.

## Chosen solution
Extend the shared unified-search request/service path with exact, root-scoped file-and-line span retrieval. Resolve paths deterministically, reject ambiguity/traversal/symlink escape, enforce a documented hard range bound, label filesystem provenance, and render identical results for CLI and MCP callers with adversarial coverage.

## Workflow notes
- Draft PR created before implementation.
- No worktree per user instruction.
- RNA MCP tools were unavailable in the agent session; required fallback exploration will be logged as friction.

## Summary by CodeRabbit

* **New Features**
  * Enabled source-span search using compiler-style `file:line:column` and optional `line`/`end_line` range filters.
  * Search can now return bounded, numbered source excerpts with repository provenance.
  * Added CLI flags to filter `search` by starting and ending lines.
* **Bug Fixes**
  * Improved propagation of `file`, `line`, `end_line`, and `root` parameters through the MCP/CLI flow.
* **Documentation**
  * Expanded README guidance on supported syntax, root scoping, and validation/limit behavior.
* **Tests**
  * Added/updated smoke tests and unit coverage for source-span handling and edge cases.

### CODE REVIEW

File: src/open_viewer.rs [hunk 1]
```diff
@@ -186,6 +186,9 @@ async fn dispatch_tool(state: &ViewerState, call: &McpCall) -> Result<String, St
                 kind,
                 language: None,
                 file: None,
+                line: None,
+                end_line: None,
+                root: None,
```
[the remark this answers is not in the document]
Comment (muness) [reply src/open_viewer.rs]: Fixed in 70081c99: the viewer adapter now forwards file, tolerant line/end_line, and root into the shared SearchParams path.

File: src/service/search.rs [hunk 2]
```diff
@@ -35,6 +38,307 @@ const IMPACT_SUMMARY_NODE_THRESHOLD: usize = 30;
...
+                    selected.push(std::mem::take(&mut current));
+                }
+                if line_number == end {
+                    completed = true;
+                    continue;
+                }
+                line_number = line_number.saturating_add(1);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (muness) [reply src/service/search.rs:108]: Fixed in 70081c99: source-span filesystem work now runs in spawn_blocking, suffix traversal is deterministically capped at 50,000 entries, and traversal-limit regression coverage was added.

File: src/service/search.rs [hunk 3]
```diff
@@ -35,6 +38,307 @@ const IMPACT_SUMMARY_NODE_THRESHOLD: usize = 30;
...
+        let Ok(canonical_root) = fs::canonicalize(&root) else {
+            continue;
+        };
+        let exact = root.join(&requested);
+        let mut paths = if exact.symlink_metadata().is_ok() {
+            vec![exact]
+        } else {
+            let mut suffix_matches = Vec::new();
+            collect_suffix_matches(&root, &requested, &mut suffix_matches);
+            suffix_matches
```
[the remark this answers is not in the document]
Comment (muness) [reply src/service/search.rs:300]: Fixed in 70081c99: suffix candidates are capped at 20 plus an omission sentinel; the diagnostic lists only the first 20 and a regression test verifies the bound.
